packages/confluence-adapter/src/get_users_and_groups.ts

```diff
@@ -0,0 +1,140 @@
+  name: Joi.string().required(),
+  id: Joi.string().required(),
+})
+  .unknown(true)
+  .required()
+
+const isUser = createSchemeGuard<User>(USER_SCHEMA)
+const isGroup = createSchemeGuard<Group>(GROUP_SCHEMA)
+
+const GROUP_FETCH_DEF: definitionsUtils.fetch.InstanceFetchApiDefinitions<Options> = {
+  requests: [
+    {
+      endpoint: {
+        method: 'get',
+        path: '/wiki/rest/api/group',
+      },
+      transformation: {
+        root: 'results',
+        pick: ['name', 'id'],
+      },
+    },
+  ],
+}
+
+const USERS_FETCH_DEF: definitionsUtils.fetch.InstanceFetchApiDefinitions<Options> = {
+  requests: [
+    {
+      endpoint: {
+        method: 'get',
+        path: '/rest/api/3/users/search',
+        queryArgs: {
+          maxResults: '10000',
+        },
+      },
+      transformation: {
+        pick: ['accountId', 'emailAddress', 'displayName'],
+      },
+    },
+  ],
+}
+
+export const getUsersAndGroups = async (
+  definitions: definitionsUtils.ApiDefinitions<Options>,
+): Promise<{ usersIndex: Record<string, User>; groupsIndex: Record<string, Group> }> => {
+  const { fetch } = definitions
+  if (fetch === undefined) {
+    throw new Error('could not find fetch definitions')
+  }
+  const createCustomizations = (): Record<string, definitionsUtils.fetch.InstanceFetchApiDefinitions<Options>> => ({
+    [GROUP_TYPE_NAME]: GROUP_FETCH_DEF,
+    [USER_TYPE_NAME]: USERS_FETCH_DEF,
+  })
+  const definitionsWithFetch = { fetch, ...definitions }
+  const requester = fetchUtils.request.getRequester<Options>({
+    adapterName: ADAPTER_NAME,
+    clients: definitionsWithFetch.clients,
+    pagination: definitionsWithFetch.pagination,
+    requestDefQuery: definitionsUtils.queryWithDefault(
+      definitionsUtils.getNestedWithDefault(
+        {
+          ...definitionsWithFetch.fetch.instances,
+          customizations: createCustomizations(),
+        },
+        'requests',
+      ),
+    ),
+  })
+  let groupResult: definitionsUtils.GeneratedItem<Record<string, unknown>, Values>[] = []
+  let userResults: definitionsUtils.GeneratedItem<Record<string, unknown>, Values>[] = []
+  let groupsIndex: Record<string, Group> = {}
+  let usersIndex: Record<string, User> = {}
+  try {
+    groupResult = await requester.requestAllForResource({
+      callerIdentifier: { typeName: GROUP_TYPE_NAME },
+      contextPossibleArgs: {},
+    })
+    groupsIndex = _.keyBy(groupResult.map(item => item.value).filter(isGroup), group => group.id)
+  } catch (e) {
+    log.error('Failed to fetch groups with error %o', e)
+    groupsIndex = {}
+  }
+  try {
+    userResults = await requester.requestAllForResource({
+      callerIdentifier: { typeName: USER_TYPE_NAME },
+      contextPossibleArgs: {},
+    })
+    usersIndex = _.keyBy(userResults.map(item => item.value).filter(isUser), user => user.accountId)
+  } catch (e) {
+    log.error('Failed to fetch users with error %o', e)
+    usersIndex = {}
+  }
+  return { groupsIndex, usersIndex }
```

Comment:
Nit (possible way to share code here):
```suggestion
const callSingleResource = async <T>(
    typeName: string,
    filterFn: (item: any) => item is T,
    idKey: string
  ): Promise<Record<string, T>> => {
    try {
      const result = await requester.requestAllForResource({
        callerIdentifier: { typeName },
        contextPossibleArgs: {},
      })
      return _.keyBy(result.map(item => item.value).filter(filterFn), idKey)
    } catch (e) {
      log.error(`Failed to fetch ${typeName.toLowerCase()}s with error %o`, e)
      return {}
    }
  }

  const [groupsIndex, usersIndex] = await Promise.all([
    callSingleResource<Group>(GROUP_TYPE_NAME, isGroup, 'id'),
    callSingleResource<User>(USER_TYPE_NAME, isUser, 'accountId'),
  ])

  return { groupsIndex, usersIndex }
```

Reply:
Nice